soccer/gameplay/tactics/wall.py

```diff
@@ -0,0 +1,96 @@
+import composite_behavior
+import behavior
+import constants
+import robocup
+import main
+import enum
+import math
+import skills.move
+
+# This tactic builds a wall a certain distance from point A, blocking point B.
+class Wall(composite_behavior.CompositeBehavior):
+    class State(enum.Enum):
+        defense_wall = 1
+
+    def __init__(self,
+                 num_defenders = 3,                         # number of defenders we're making the wall with (default 3)
+                 curvature = 0,                             # curvature (in radians) of the wall 0 - pi/2
+                 mark_point = None,                         # what point we are defending against (default is ball)
+                 defender_point = robocup.Point(0, 0),      # what point we are defending (default is goal)
+                 defender_spacing = 3,                      # number of robot radii between the centers of the defenders in the wall
+                 dist_from_mark = 1):                       # distance from the mark point we want to build the wall
+        super().__init__(continuous=True)
+
+        self.number_of_defenders = num_defenders
+        self.curvature = curvature
+        self._mark_point = main.ball().pos if mark_point == None else mark_point
+        self._defense_point = defender_point
+        self.dist_from_mark = dist_from_mark
+        self.defender_spacing = defender_spacing
+
+        self.add_state(Wall.State.defense_wall, 
+                       behavior.Behavior.State.running)
+
+        self.add_transition(behavior.Behavior.State.start,
+                            Wall.State.defense_wall, lambda: True,
+                            "immediately")
+
+        self.add_transition(
+            Wall.State.defense_wall,
+            behavior.Behavior.State.completed, lambda: self.completed(), "robot kicked")
```

Comment:
Make sure to also have a transition from completed to defense_wall again if one of the robots get's knocked out of the way or if we end up changing the mark point after all the robots have hit their target locations

Reply:
does it make sense to even have a completed state?

Reply:
The only case it would be useful is if the parent behavior wants to know when all the robots are in their specified locations. Other than that, it makes no difference at all. If you would use that information in the defense, go for it and add it in, otherwise it doesn't really make a difference